ai_edge_torch/generative/layers/unet/blocks_2d.py

```diff
@@ -0,0 +1,287 @@
+      x = x + time_emb
+    x = self.norm_2(x)
+    x = self.act_fn(x)
+    x = self.conv_2(x)
+    x = x + self.residual_layer(residual)
+    return x
+
+
+class AttentionBlock2D(nn.Module):
+  """2D self attention block
+
+  x = SelfAttention(Norm(input_tensor))
+
+  """
+
+  def __init__(self, config: unet_cfg.AttentionBlock2DConfig):
+    """Initialize an instance of the AttentionBlock2D.
+
+    Args:
+      config (unet_cfg.AttentionBlock2DConfig): the configuration of this block.
+    """
+    super().__init__()
+    self.norm = layers_builder.build_norm(config.dims, config.normalization_config)
+    self.attention = SelfAttention(config.dims, config.attention_config, 0, True)
+
+  def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+    """Forward function of the AttentionBlock2D.
+
+    Args:
+      input_tensor (torch.Tensor): the input tensor.
+
+    Returns:
+      output activation tensor after self attention.
+    """
+    residual = input_tensor
+    x = self.norm(input_tensor)
+    B, C, H, W = x.shape
+    x = x.view(B, C, H * W)
+    x = x.transpose(-1, -2)
+    x = self.attention(x)
+    x = x.transpose(-1, -2)
+    x = x.view(B, C, H, W)
+    x = x + residual
+    return x
+
+
+class UpDecoderBlock2D(nn.Module):
+  """Decoder block containing several residual blocks followed by an optional upsampler.
+
+       input_tensor
+            |
+            ▼
+  ┌───────────────────┐
+  │  ResidualBlock2D  │ num_layers
+  └─────────┬─────────┘
+            │
+  ┌─────────▼─────────┐
+  │    (Optional)     │
+  │     Upsampler     │
+  └─────────┬─────────┘
+            │
+  ┌─────────▼─────────┐
+  │    (Optional)     │
+  │      Conv2D       │
+  └─────────┬─────────┘
+            │
+            ▼
+      hidden_states
+  """
+
+  def __init__(self, config: unet_cfg.UpDecoderBlock2DConfig):
+    """Initialize an instance of the UpDecoderBlock2D.
+
+    Args:
+      config (unet_cfg.UpDecoderBlock2DConfig): the configuration of this block.
+    """
+    super().__init__()
+    self.config = config
+    resnets = []
+    for i in range(config.num_layers):
+      input_channels = config.in_channels if i == 0 else config.out_channels
+      resnets.append(
+          ResidualBlock2D(
+              unet_cfg.ResidualBlock2DConfig(
+                  in_channels=input_channels,
+                  out_channels=config.out_channels,
+                  time_embedding_channels=config.time_embedding_channels,
+                  normalization_config=config.normalization_config,
+                  activation_type=config.activation_type,
+              )
+          )
+      )
+    self.resnets = nn.ModuleList(resnets)
+    if config.add_upsample:
+      self.upsampler = unet_builder.build_upsampling(config.sampling_config)
+      if config.upsample_conv:
+        self.upsample_conv = nn.Conv2d(
+            config.out_channels, config.out_channels, kernel_size=3, stride=1, padding=1
```

Comment:
are kernel sizes configurable?

Reply:
Could be configurable. But for current use case, UNet and AutoEncoder,  upsampling Conv all use kernel size 3.
Try not to add too many config options one time. We can extend the configs when needed.

Reply:
yeah for sure.